### PR TITLE
:sparkles: Add support for github GHES

### LINKED ISSUE #660: Support private git servers (like Github Enterprise instances)
**Is your feature request related to a problem? Please describe.**
I would like to run scorecard for a repository that is hosted on my [Github Enterprise](https://enterprise.github.com/) instance. I currently can't.

```
$ scorecard --repo github.corp.net/nv35/myrepo
2021/07/05 09:24:35 unsupported host: github.corp.net
```

**Describe the solution you'd like**
Allow to use the others github instances just like the .com one. The API path would likely just change from https://api.github.com/ to https://github.corp.net/api/v3, and I guess the calls would just be the same.

**Additional context**
The restriction to github.com is defined here: https://github.com/ossf/scorecard/blob/1f1e05b22ccfac2c389125cdbc5be0bc61a52c1e/repos/repo_url.go#L77

### PR DESCRIPTION
**Pull request by patelniketm**
#### What kind of change does this PR introduce?

Support GitHub GHES by using Enterprise Client for the GitHub Calls

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

```
$ scorecard --repo github.corp.com/org/repo
unsupported host: github.corp.com
```

#### What is the new behavior (if this is a feature change)?**

```
$ GH_HOST=github.corp.com scorecard --repo github.corp.com/org/repo
```

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Fixes #660 

#### Special notes for your reviewer

This is based of stale #2788 so credit to author. Locally, I was able to validate this with our GHES instance. Please let me know if I'm missing any coverage or any expectations here.

#### Does this PR introduce a user-facing change?

No changes needed for existing users. This is additional feature to support GHES, for those users release note updated. To keep consistency with `gh cli`, using same env variable name to keep consistency.

```release-note
To use GitHub Enterprise Server (GHES),
Please define `GH_HOST` env variable, such as `github.corp.com` or `corp.github.com`
```

### CODE REVIEW

**Comment by patelniketm** [comment]
Locally on Mac - golang 1.20.4 and 1.19 version, all tests pass successfully and consistently. I'm suspecting cache in the pipeline might be causing this. I would attempt to run this on GHA ubuntu runner to see if that is the issue. 

If you have any suggestions or input, please let me know.

File: clients/githubrepo/client.go [hunk 1]
```diff
@@ -259,8 +267,26 @@ func CreateGithubRepoClientWithTransport(ctx context.Context, rt http.RoundTripp
 	httpClient := &http.Client{
 		Transport: rt,
 	}
-	client := github.NewClient(httpClient)
-	graphClient := githubv4.NewClient(httpClient)
+
+	var client *github.Client
+	var graphClient *githubv4.Client
+	githubHost, isGhHost := os.LookupEnv("GH_HOST")
```
Comment (raghavkaul) [inline clients/githubrepo/client.go:273]: Could you add documentation indicating when this environment variable should be set?

File: clients/githubrepo/repo.go [hunk 2]
```diff
@@ -83,8 +88,10 @@ func (r *repoURL) String() string {
 
 // IsValid implements Repo.IsValid.
 func (r *repoURL) IsValid() error {
+	githubHost := os.Getenv("GH_HOST")
```
Comment (raghavkaul) [inline clients/githubrepo/repo.go:91]: Would a user be setting the GH_HOST field twice? Could they instead pass in `--repo-url` and we use the environment variable to determine whether GH Enterprise is being used, or not?

File: clients/githubrepo/repo_test.go [hunk 3]
```diff
@@ -78,11 +81,25 @@ func TestRepoURL_IsValid(t *testing.T) {
...
+			wantErr:  false,
+			ghHost:   true,
+		},
 	}
 	for _, tt := range tests {
 		tt := tt // Re-initializing variable so it is not changed while executing the closure below
 		t.Run(tt.name, func(t *testing.T) {
-			t.Parallel()
+			if tt.ghHost {
+				os.Setenv("GH_HOST", "github.corp.com")
```
Comment (raghavkaul) [inline clients/githubrepo/repo_test.go:100]: Should unset this after the test is over (see: https://beta.pkg.go.dev/testing#T.Cleanup).

**Comment by raghavkaul** [comment]
Thank you for the contribution! Left some comments about the approach. Would be good to get to the bottom of the test failures, since those are requirements for merging. You may have environment variables set on your local tests that GitHub CI doesn't set automatically.

File: clients/githubrepo/repo.go (hunk 2 above)
Comment (patelniketm) [reply clients/githubrepo/repo.go:91]: It is intended to setup - GH_HOST only once as environment variable similar to GITHUB_TOKEN. Though, consumer might configure github host twice (env and repo full URL). We can clarify this in documentation if that helps. In short, we only need one and env version fit their.

`GH_HOST` is one of [github cli environment](https://cli.github.com/manual/gh_help_environment) based approach to switch GHES and provide convenience.

Technically we can use without env version and push consumer to use full URL to identify whether it is GHEC or GHES, though Gitlab in mix can add conditional check. IDK if GitLab also have its Cloud and Enterprise version. With those uncertainty I see GH_HOST as env fit well.

Though I'm open on either way.

File: clients/githubrepo/repo_test.go (hunk 3 above)
Comment (patelniketm) [reply clients/githubrepo/repo_test.go]: To keep things consistent with other test, addressing this with `t.Setenv` which auto cleanup

File: clients/githubrepo/client.go (hunk 1 above)
Comment (patelniketm) [reply clients/githubrepo/client.go:273]: I have added [documentation](https://github.com/ossf/scorecard/pull/2999/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R407) as requested.

**Comment by patelniketm** [comment]
@raghavkaul Thanks for your input on environment, that was the reason for failing test. I have fixed the test.

File: clients/githubrepo/repo.go (hunk 2 above)
Comment (raghavkaul) [reply clients/githubrepo/repo.go:91]: Gotcha, I wasn't aware that it was also a GitHub CLI environment variable. That makes sense.

**Review by raghavkaul — APPROVED** [review]

**Pull request merged**